File: src/app/special-layers.ts
import type { Animation } from '../domain/types';

export interface SpecialLayerIndices {
  plantCustomLayers: number[];
  zombieStateLayers: number[];
  groundSwatchLayers: number[];
  defaultHiddenLayers: number[];
}

const LOWERCASE_EXACT_DEFAULT_HIDDEN_LAYERS = new Set([
  'magnet_item',
]);

const SHADOW_POWER_ANIMATION_STEMS = [
  'dragonbabybruit',
  'dragonbruit',
  'dusklobber',
  'gloomvine',
  'grimrose',
  'guardshroom',
  'moonflower',
  'murkadamia',
  'nightshade',
  'noctarine',
  'powervine',
  'shadowpea',
  'shadowshroom',
];

export function getSpecialLayerIndices(anim: Animation, sourceName = ''): SpecialLayerIndices {
  const result: SpecialLayerIndices = {
    plantCustomLayers: [],
    zombieStateLayers: [],
    groundSwatchLayers: [],
    defaultHiddenLayers: [],
  };
  const sourceStem = getSourceStem(sourceName);
  const sourceKey = normalizeSourceKey(sourceStem);
  const isNightshadeAnimation = sourceKey.includes('nightshade');
  const isShadowPowerAnimation = isShadowPowerSource(sourceKey);
  const spriteNames = new Set(anim.sprite.map(sp => sp.name).filter((name): name is string => !!name));

  anim.sprite.forEach((sp, i) => {
    if (!sp.name) return;
    const name = sp.name;
    const lowerName = name.toLowerCase();
    if (lowerName.startsWith('custom_')) result.plantCustomLayers.push(i);
    if (name === 'ink' || name === 'butter') result.zombieStateLayers.push(i);
    if (name === 'ground_swatch' || name === 'ground_swatch_plane') result.groundSwatchLayers.push(i);
    if (isDefaultHiddenSpriteName(name, isNightshadeAnimation, isShadowPowerAnimation, spriteNames)) {
      result.defaultHiddenLayers.push(i);
    }
  });

  return result;
}

function getSourceStem(sourceName: string): string {
  const baseName = sourceName.split(/[\\/]/).pop() ?? sourceName;
  return baseName
    .replace(/\.pam\.json$/i, '')
    .replace(/\.pam\.ya?ml$/i, '')
    .replace(/\.pam\.toml$/i, '')
    .replace(/\.json$/i, '')
    .replace(/\.ya?ml$/i, '')
    .replace(/\.toml$/i, '')
    .replace(/\.pam$/i, '')
    .replace(/\.fla$/i, '');
}

function normalizeSourceKey(sourceStem: string): string {
  return sourceStem.toLowerCase().replace(/[^a-z0-9]+/g, '');
}

function isShadowPowerSource(sourceKey: string): boolean {
  return SHADOW_POWER_ANIMATION_STEMS.some(stem => sourceKey.includes(stem));
}

function isDefaultHiddenSpriteName(
  name: string,
  isNightshadeAnimation: boolean,
  isShadowPowerAnimation: boolean,
  spriteNames: ReadonlySet<string>,
): boolean {
  const lowerName = name.toLowerCase();
  return lowerName.startsWith('custom_')
    || lowerName.includes('armor')
    || LOWERCASE_EXACT_DEFAULT_HIDDEN_LAYERS.has(lowerName)
    || isDarkLayerName(name, isShadowPowerAnimation, spriteNames)
    || (isNightshadeAnimation && lowerName.endsWith('_pf'));
}

function isDarkLayerName(
  name: string,
  isShadowPowerAnimation: boolean,
  spriteNames: ReadonlySet<string>,
): boolean {
  return isPairedDarkLayer(name, spriteNames)
    || (isShadowPowerAnimation && name.toLowerCase().includes('dark'));
}

function isPairedDarkLayer(name: string, spriteNames: ReadonlySet<string>): boolean {
  if (name.endsWith('_dark')) {
    const base = name.slice(0, -'_dark'.length);
    if (spriteNames.has(base) || spriteNames.has(`${base}_normal`) || spriteNames.has(`${base}_white`)) {
      return true;
    }
  }

  const darkInfix = '_dark_';
  const infixIndex = name.indexOf(darkInfix);
  if (infixIndex !== -1) {
    const lightName = name.slice(0, infixIndex) + '_' + name.slice(infixIndex + darkInfix.length);
    if (spriteNames.has(lightName)) return true;
  }

  const darkPrefix = 'dark_';
  if (name.startsWith(darkPrefix)) {
    if (spriteNames.has(name.slice(darkPrefix.length))) return true;
  }

  return false;
}

File: crates/core/src/special_layers.rs
use std::collections::HashSet;

use pam_codec::PamInfo;

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct SpecialLayerIndices {
    pub plant_custom_layers: Vec<usize>,
    pub zombie_state_layers: Vec<usize>,
    pub ground_swatch_layers: Vec<usize>,
    pub default_hidden_layers: Vec<usize>,
}

const SHADOW_POWER_STEMS: &[&str] = &[
    "dragonbabybruit",
    "dragonbruit",
    "dusklobber",
    "gloomvine",
    "grimrose",
    "guardshroom",
    "moonflower",
    "murkadamia",
    "nightshade",
    "noctarine",
    "powervine",
    "shadowpea",
    "shadowshroom",
];

pub fn special_layer_indices(pam: &PamInfo, source_name: &str) -> SpecialLayerIndices {
    let mut result = SpecialLayerIndices::default();
    let source_key = source_stem(source_name)
        .chars()
        .filter(|ch| ch.is_ascii_alphanumeric())
        .flat_map(char::to_lowercase)
        .collect::<String>();
    let nightshade = source_key.contains("nightshade");
    let shadow_power = SHADOW_POWER_STEMS
        .iter()
        .any(|stem| source_key.contains(stem));
    let names = pam
        .sprite
        .iter()
        .filter_map(|sprite| sprite.name.as_deref())
        .collect::<HashSet<_>>();

    for (index, sprite) in pam.sprite.iter().enumerate() {
        let Some(name) = sprite.name.as_deref() else {
            continue;
        };
        let lower = name.to_ascii_lowercase();
        if lower.starts_with("custom_") {
            result.plant_custom_layers.push(index);
        }
        if matches!(lower.as_str(), "ink" | "butter") {
            result.zombie_state_layers.push(index);
        }
        if matches!(lower.as_str(), "ground_swatch" | "ground_swatch_plane") {
            result.ground_swatch_layers.push(index);
        }
        if is_default_hidden(name, &lower, nightshade, shadow_power, &names) {
            result.default_hidden_layers.push(index);
        }
    }
    result
}

fn source_stem(source_name: &str) -> String {
    let file = source_name
        .rsplit(['/', '\\'])
        .next()
        .unwrap_or(source_name);
    let lower = file.to_ascii_lowercase();
    for extension in [
        ".pam.json",
        ".pam.yaml",
        ".pam.yml",
        ".pam.toml",
        ".json",
        ".yaml",
        ".yml",
        ".toml",
        ".pam",
        ".fla",
    ] {
        if lower.ends_with(extension) {
            return file[..file.len() - extension.len()].to_string();
        }
    }
    file.to_string()
}

fn is_default_hidden(
    name: &str,
    lower: &str,
    nightshade: bool,
    shadow_power: bool,
    names: &HashSet<&str>,
) -> bool {
    lower.starts_with("custom_")
        || lower.contains("armor")
        || lower == "magnet_item"
        || is_dark_layer(name, lower, shadow_power, names)
        || (nightshade && lower.ends_with("_pf"))
}

fn is_dark_layer(name: &str, lower: &str, shadow_power: bool, names: &HashSet<&str>) -> bool {
    if shadow_power && lower.contains("dark") {
        return true;
    }
    if let Some(base) = name.strip_suffix("_dark")
        && (names.contains(base)
            || names.contains(format!("{base}_normal").as_str())
            || names.contains(format!("{base}_white").as_str()))
    {
        return true;
    }
    if let Some((prefix, suffix)) = name.split_once("_dark_")
        && names.contains(format!("{prefix}_{suffix}").as_str())
    {
        return true;
    }
    if let Some(base) = name.strip_prefix("dark_")
        && names.contains(base)
    {
        return true;
    }
    false
}
